### PR TITLE
implement irregular trigger

### PR DESCRIPTION
**Pull request by naoto0804**
Implement `IrregularTrigger` (I’m worried a bit if it is correct naming) that fires when the number of epochs or iterations reaches arbitrary values in contrast to IntervalTrigger. It can accept multiple values. The typical use case is setting complex and irregular learning rate decay schedule (e.g., lr_policy: multistep mode in Caffe solver) as a trigger of `ExponentialShift`.

### CODE REVIEW

**Comment by delta2323** [comment]
I found the description is same as that of `IntervalTrigger`, could you clarify the difference of the two?

**Comment by mattya** [comment]
Thank you for your PR!

In the case of naming, I think ```Irregular``` is a little bit ambiguous, so how about ```SpecificTimeTrigger``` or like that.

Also, please add this trigger into https://github.com/pfnet/chainer/blob/master/docs/source/reference/triggers.rst
for generating docs.

File: chainer/training/triggers/irregular_trigger.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+class IrregularTrigger(object):
+
+    """Trigger invoked when specific value reaches the indicated values.
```
Comment (mattya) [inline chainer/training/triggers/irregular_trigger.py]: Currently this trigger can specify only `iteration` or `epoch`.  So 
```Trigger invoked at specified iteration or epoch.```
is better to understand.

**Comment by beam2d** [comment]
Another naming idea: `ManualScheduleTrigger`.

File: chainer/training/triggers/irregular_trigger.py (hunk 1 above)
Comment (naoto0804) [inline chainer/training/triggers/irregular_trigger.py]: That's right, thanks! I will rewrite the doc more exactly.

**Comment by naoto0804** [comment]
I decided to use `ManualScheduleTrigger`. Thank you for your ideas!
I have fixed and rewrote the doc more exactly.

File: chainer/training/triggers/manual_schedule_trigger.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        unit (str): Unit of the time specified by ``points``. It must be
+            either ``'iteration'`` or ``'epoch'``.
+
+    """
+
+    def __init__(self, points, unit):
+        self.points = points if isinstance(points, list) else [points]
+        assert unit == 'epoch' or unit == 'iteration'
+        self.unit = unit
+        self.count = [0 for _ in range(len(self.points))]
```
Comment (unnonouno) [inline chainer/training/triggers/manual_schedule_trigger.py]: `[0] * len(self.points)`?

File: chainer/training/triggers/manual_schedule_trigger.py [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+        self.valid = [True for _ in range(len(self.points))]
```
Comment (unnonouno) [inline chainer/training/triggers/manual_schedule_trigger.py]: `[True] * len(self.points)`?

File: chainer/training/triggers/manual_schedule_trigger.py [hunk 4]
```diff
@@ -0,0 +1,50 @@
+class ManualScheduleTrigger(object):
+
+    """Trigger invoked at specified point(s) of iterations or epochs.
+
+    This trigger accepts iterations or epochs indicated by given point(s).
+    There are two ways to specify the point(s): iteration and epoch.
+    `iteration` means the number of updates, while `epoch` means the number
+    of sweeps over the training dataset. Fractional values are allowed
+    if the point is a number of epochs; the trigger uses the `iteration`
+    and `epoch_detail` attributes defined by the updater.
```
Comment (unnonouno) [inline chainer/training/triggers/manual_schedule_trigger.py]: Use double backquotes for code snippets.

File: chainer/training/triggers/manual_schedule_trigger.py [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+        Returns:
+            bool: True if the corresponding extension should be invoked in this
+                iteration.
+
+        """
+        updater = trainer.updater
+        if self.unit == 'epoch':
+            prev = self.count
+            self.count = [updater.epoch_detail // p for p in self.points]
+            flag = [p != c for p, c in zip(prev, self.count)]
```
Comment (unnonouno) [inline chainer/training/triggers/manual_schedule_trigger.py]: Use `six.moves.zipe`

File: chainer/training/triggers/manual_schedule_trigger.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+
+        Returns:
+            bool: True if the corresponding extension should be invoked in this
+                iteration.
+
+        """
+        updater = trainer.updater
+        if self.unit == 'epoch':
+            prev = self.count
+            self.count = [updater.epoch_detail // p for p in self.points]
```
Comment (mattya) [inline chainer/training/triggers/manual_schedule_trigger.py]: If you resume the learning from a snapshot, the epoch can start from not 1, but ManualScheduleTrigger([1], 'epoch') will fire at the first iteration.
I think this is an undesired behaviour.

**Comment by naoto0804** [comment]
Oops.. Thanks for your comments. 
I've found that using `updater.epoch_detail // p` and `self.count` is redundant and even causes a problem that you kindly pointed out.
I have modified the code to be simpler and clearer without using them and reflect your comments.

File: chainer/training/triggers/manual_schedule_trigger.py (hunk 6 above)
Comment (naoto0804) [inline chainer/training/triggers/manual_schedule_trigger.py]: I've found that using updater.epoch_detail // p and self.count is redundant.
I've modified.

**Comment by mattya** [comment]
Thanks for the update but it seems the problem is not yet fixed.
For example, if p=3 (epoch) and resume the learning from epoch 10,
both `valid` and `updater.epoch_detail >= p` will be satisfied at the first iteration.
How about also check `p > updater.epoch_detail - (epoch per iteration)`?

**Comment by naoto0804** [comment]
Sorry for my misunderstanding.
I've checked `updater.epoch_detail < p + (epoch per iteration)` and handle exception of the first iteration in resuming training.

**Comment by naoto0804** [comment]
@mattya 
Sorry for my misunderstanding.
I've checked `updater.epoch_detail < p + (epoch per iteration)` and handle exception of the first iteration in resuming training.

**Comment by Hakuyume** [comment]
I've added test cases for resumed trainer.

File: chainer/training/triggers/manual_schedule_trigger.py [hunk 7]
```diff
@@ -0,0 +1,54 @@
+#!/usr/bin/python
```
Comment (okuta) [inline chainer/training/triggers/manual_schedule_trigger.py]: Please remove this line.

File: chainer/training/triggers/manual_schedule_trigger.py [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+# -*- coding: utf-8 -*-
```
Comment (okuta) [inline chainer/training/triggers/manual_schedule_trigger.py]: Please remove this line.

**Comment by Hakuyume** [comment]
I modified `ManualScheduleTrigger` to use `previous_epoch_detail` property. This modification solves resuming issue which @mattya mentioned at https://github.com/pfnet/chainer/pull/2181#issuecomment-277150617. This is the same solution as #2484 .
